server/repository/src/migrations/mod.rs

```diff
@@ -0,0 +1,156 @@
+mod types;
+mod v1_00_04;
+mod version;
+pub(crate) use self::types::*;
+use self::v1_00_04::V1_00_04;
+
+mod templates;
+
+pub(crate) use self::version::*;
+
+use crate::{
+    run_db_migrations, DBConnection, DBType, KeyValueStoreRepository, KeyValueType,
+    RepositoryError, StorageConnection,
+};
+use diesel::{query_builder::QueryFragment, query_dsl::methods, sql_query, RunQueryDsl};
+use thiserror::Error;
+
+pub(crate) trait Migration {
+    fn version(&self) -> Version;
+    fn migrate(&self, connection: &StorageConnection) -> anyhow::Result<()>;
+}
+
+#[derive(Debug, Error)]
+pub enum MigrationError {
+    #[error("Database version ({0}) is above app version ({1})")]
+    DatabaseVersionAboveAppVersion(Version, Version),
+    #[error("Database version is pre release ({0}), it cannot be upgraded")]
+    DatabaseVersionIsPreRelease(Version),
```

Comment:
There is a small not about this in readme.md

Reply:
not(about this) 😆

Reply:
Is is possible to have a nicer error when you have the database version out of sync?
Currently this is just a panic and the server fails to start. The message is clear if you know what to look for, but could be clearer with a few words around it.. `The database you are connecting to is a later version (1.1.1) than the server (1.1.0). It is unsafe to run with this configuration, the server is stopping.`

🤷 

```
2022-11-30 10:43:32.721533 INFO repository -
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Failed to run DB migrations

Caused by:
    Database version (1.1.1) is above app version (1.1.0)', /Users/m2/GitHub/open-msupply/server/server/src/lib.rs:70:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Reply:
I agree that's much better, 95cd0b35a0a4675b97ab8446c96a3ee351379225